src/hip_memory.cpp

```diff
@@ -1540,111 +1540,136 @@ hipError_t hipMemcpyAtoH(void* dst, hipArray* srcArray, size_t srcOffset, size_t
     return ihipLogStatus(e);
 }
 
+int getByteSizeFromFormat(const hipChannelFormatDesc& desc){
+    int byteSize =0;
+    switch (desc.f) {
+        case hipChannelFormatKindUnsigned:
+            switch (desc.x) {
+                case 32:
+                    byteSize = sizeof(uint32_t);
+                    break;
+                case 16:
+                    byteSize = sizeof(uint16_t);
+                    break;
+                case 8:
+                    byteSize = sizeof(uint8_t);
+                    break;
+                default:
+                    byteSize = sizeof(uint32_t);
+            }
+            break;
+        case hipChannelFormatKindSigned:
+            switch (desc.x) {
+                case 32:
+                    byteSize = sizeof(int32_t);
+                    break;
+                case 16:
+                    byteSize = sizeof(int16_t);
+                    break;
+                case 8:
+                    byteSize = sizeof(int8_t);
+                    break;
+                default:
+                    byteSize = sizeof(int32_t);
+            }
+            break;
+        case hipChannelFormatKindFloat:
+            switch (desc.x) {
+                case 32:
+                    byteSize = sizeof(float);
+                    break;
+                case 16:
+                    byteSize = sizeof(_Float16);
+                    break;
+                default:
+                    byteSize = sizeof(float);
+            }
+            break;
+        case hipChannelFormatKindNone:
+        default:
+            break;
+    }
+    return byteSize;
+}
+
 hipError_t ihipMemcpy3D(const struct hipMemcpy3DParms* p, hipStream_t stream, bool isAsync) {
     hipError_t e = hipSuccess;
     if(p) {
-        size_t byteSize, width, height, depth, widthInBytes, srcPitch, dstPitch, ySize;
-        hipChannelFormatDesc desc;
+        size_t dstByteSize, srcByteSize, copyWidth, copyHeight, copyDepth, widthInBytes, srcPitch, dstPitch, srcYsize, dstYsize;
+        size_t srcXoffset, srcYoffset, srcZoffset, dstXoffset, dstYoffset, dstZoffset;
+        size_t srcWidth, srcHeight, srcDepth, dstWidth, dstHeight, dstDepth;
+
         void* srcPtr;void* dstPtr;
+        copyDepth = p->extent.depth;
+        copyHeight = p->extent.height;
+        copyWidth =  p->extent.width; // in bytes ?
+        dstXoffset = p->dstPos.x;
+        dstYoffset = p->dstPos.y;
+        dstZoffset = p->dstPos.z;
+        srcXoffset = p->srcPos.x;
+        srcYoffset = p->srcPos.y;
+        srcZoffset = p->srcPos.z;
```

Comment:
Here and below: I'd think about hipMemcpy3DParms struct var on the stack for copying routines.